org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/ProjectDependenciesTest.scala

```diff
@@ -14,40 +14,87 @@ import org.eclipse.jdt.core.IPackageFragment
 import org.scalaide.util.internal.SettingConverterUtil
 import org.scalaide.ui.internal.preferences.ScalaPluginSettings
 import org.scalaide.core.IScalaPlugin
+import org.eclipse.core.resources.IProject
 
 class ProjectDependenciesTest {
 
   import SDTTestUtils._
 
+  private implicit class TestableProject(from: IScalaProject) {
+    def dependsOnAndExports(dep: IScalaProject, exported: Boolean = true): Unit =
+      addToClasspath(from, JavaCore.newProjectEntry(dep.underlying.getFullPath, exported))
+
+    def onlyDependsOn(dep: IScalaProject) = dependsOnAndExports(dep, false)
+
+    def shouldDependsOn(msg: String, deps: IScalaProject*): Unit = {
+      val expected = deps.map(_.underlying).sortBy(_.getName)
+      val computed = from.transitiveDependencies.sortBy(_.getName)
+      Assert.assertEquals(s"${msg.capitalize} for ${from.underlying.getName}", expected, computed)
+    }
+  }
+
+
+  @Test def transitive_dependencies_more_complicated_tree(): Unit = {
+    val allProj @ Seq(prjA, prjB, prjC, prjD, prjE, prjF, prjG) = createProjects("A", "B", "C", "D", "E", "F", "G")
+    try {
+      prjB dependsOnAndExports prjA
+
+      prjD dependsOnAndExports prjC
+
+      prjE onlyDependsOn prjB
+
+      prjE dependsOnAndExports prjD
+
+      prjF dependsOnAndExports prjA
+      prjF dependsOnAndExports prjE
+
+      prjG dependsOnAndExports prjE
+      prjG dependsOnAndExports prjC
+
+
+      prjB.shouldDependsOn("Only A", prjA)
+      prjD.shouldDependsOn("Only C", prjC)
+      prjE.shouldDependsOn("A, B, C,D", prjB, prjD, prjA, prjC)
+      prjG.shouldDependsOn("C, D, E - C should not be added twice, B and A should be excluded", prjC, prjE, prjD)
+
+    } finally {
```

Comment:
You could move the `deleteProjects` to a method annotation with `@After` and make `allProj` a class value.